CMakeLists.txt

```diff
@@ -318,11 +318,92 @@ include(cmake/ssl.cmake)
 include(cmake/upb.cmake)
 include(cmake/xxhash.cmake)
 include(cmake/zlib.cmake)
+include(cmake/download_archive.cmake)
+
+# Setup external proto library at third_party/envoy-api with 2 download URLs
+if (NOT EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/third_party/envoy-api)
+  # Download the archive via HTTP, validate the checksum, and extract to third_party/envoy-api.
+  download_archive(
```

Comment:
note that this basically kills the idea of having offline cmake builds (which might or might not be a good idea, I'll have to give that a thought).

Reply:
I see 2 alternatives here:

1. Make all dependencies here sub-tree (like what we did for upb, xxhash: https://github.com/grpc/grpc/tree/master/third_party);
2. Stop supporting C++ xDS related targets in CMake (CSDS, Admin, and several tests).

WDYT?

Reply:
Let's use the approach suggested in this PR.
Btw, it seems that if you pre-populate the right directories (e.g. `third_party/envoy-api`) with what it should contains, there will be no need to perform a download.